src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -105,6 +107,41 @@ public static unsafe string GetAsciiStringNonNullCharacters(this Span<byte> span
             return asciiString;
         }
 
+        public static unsafe string GetAsciiOrUTF8StringNonNullCharacters(this Span<byte> span)
+        {
+            if (span.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            var resultString = new string('\0', span.Length);
+
+            fixed (char* output = resultString)
+            fixed (byte* buffer = &MemoryMarshal.GetReference(span))
+            {
+                // This version if AsciiUtilities returns null if there are any null (0 byte) characters
+                // in the string
+                if (!StringUtilities.TryGetAsciiString(buffer, output, span.Length))
+                {
+                    // null characters are considered invalid
+                    if (span.IndexOf((byte)0) != -1)
+                    {
+                        throw new InvalidOperationException();
```

Comment:
Are these the normal types of exceptions we'd throw here? No messages? (Same question about the exception below.)

Reply:
I was a little surprised as well but it is what we do currently: https://github.com/aspnet/KestrelHttpServer/blob/release/2.2/src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs#L100-L103.

Reply:
@Eilon You're right that this is a really bad idea, but as @JunTaoLuo mentions, it's the pattern we currently use. This exception gets caught and handled, but the fact that it's a generic InvalidOperationException certainly creates the concern that we might swallow other exceptions unintentionally.

Last month, I created #2737 to track work on fixing this. It's currently in the 3.0 milestone.